src/chrome/internal/stackTraces/callFrame.ts

```diff
@@ -0,0 +1,129 @@
+import { ILoadedSource } from '../sources/loadedSource';
+import { IScript } from '../scripts/script';
+import { Protocol as CDTP } from 'devtools-protocol';
+import { ICallFrameFunction } from './callFrameFunction';
+import { Scope } from './scopes';
+
+/** CDTP has two types of stack traces:
+ * 1. CDTP.Runtime stack traces have only information about which code was executed, but not the state associated with it
+ * 2. CDTP.Debugger stack traces which have all the information that CDTP.Runtime has, and it also includes state information
+ *
+ * We represent this by modeling the information that both contain in CodeFlowFrame and the information that only CDTP.Debugger
+ * contains on CallFrame (which has an embedded CodeFlowFrame)
+ */
+
+export type ScriptOrLoadedSource = IScript | ILoadedSource;
+
+/** This class represents the code flow (which code was executed) of a call frame.
+ * (This has similar properties as the stack traces from the CDTP.Runtime domain)
+ */
+export class CodeFlowFrame<TResource extends ScriptOrLoadedSource> {
+    constructor(
+        public readonly index: number,
+        public readonly callFrameFunction: ICallFrameFunction,
+        public readonly location: Location<TResource>) { }
+
+    public get source(): TResource extends ILoadedSource ? TResource : never {
+        return this.location.resource as any;
+    }
+
+    public get script(): TResource extends IScript ? TResource : never {
+        return this.location.resource as any;
+    }
+
+    public get lineNumber(): number {
+        return this.location.position.lineNumber;
+    }
+
+    public get columnNumber(): number {
+        return this.location.position.columnNumber;
+    }
+
+    public get functionDescription(): string {
+        return this.callFrameFunction.description;
+    }
+}
+
+/** This interface represents both the code flow and the state of a call frame.
+ * (This has similar properties as the stack traces from the CDTP.Debugger domain)
+ */
+export interface ICallFrame<TResource extends ScriptOrLoadedSource> {
+    readonly index: number;
+    readonly location: Location<TResource>;
+    readonly lineNumber: number;
+    readonly columnNumber: number;
+    readonly codeFlow: CodeFlowFrame<TResource>;
+    readonly scopeChain: Scope[];
+    readonly frameThis?: CDTP.Runtime.RemoteObject;
+    readonly returnValue?: CDTP.Runtime.RemoteObject;
+}
+
+export type CallFrame<TResource extends ScriptOrLoadedSource> =
+    TResource extends ILoadedSource ? LoadedSourceCallFrame :
+    TResource extends IScript ? ScriptCallFrame :
+    ICallFrame<never>; // TODO: Figure out how to change this for never
+
+abstract class CallFrameCommonLogic<TResource extends ScriptOrLoadedSource> implements ICallFrame<TResource> {
+    public abstract get scopeChain(): Scope[];
+    public abstract get codeFlow(): CodeFlowFrame<TResource>;
+
+    public get source(): TResource extends ILoadedSource ? TResource : never {
```

Comment:
Could you please give me more details about what makes this confusing?

Reply:
If I am looking at properties available on the object, I see that there is `source` and also that there is a `codeFlow` which has a `source` inside it, and it's not obvious what the difference is, or whether they are the same, and which one I should use.

Reply:
If I added some comments that explained what each property is, and clearly say that both sources are the same, would that address your concern?

Reply:
Not really, why do we need two ways to access the same info?

Reply:
I think that both methods/properties make sense. I might need the source associated with a frame, or the code flow. When I need the source I'll naturally do frame.source rather than try to remember which instance variable has the source. It also avoids witting frame.codeFlow.source several times in the code.

Another good example (Although I'm going to remove this after the html scripts update) is script.url which is defined as script.runtimeSource.identifier.textRepresentation. If I want to get the url of a script it made sense in our older model to do script.url rather than have to remember to do: script.runtimeSource.identifier.textRepresentation.
Also when I read script.runtimeSource.identifier.textRepresentation it takes me longer to understand what is happening compared to just reading script.url.

I think a lot of times it's useful to have different ways to access the same information.

Reply:
Can the `codeFlow` be private?

Reply:
I've just checked that and we use it indirectly on the FramePresentationCommonLogic. Currently we access the properties like:

`
framePresentation.source and
framePresentation.lineNumber
`
And I also have some logic that makes it so that both CodeFlow and CallFrame use the same interface (which uses the public codeflow property).

If I remove all the "duplicated sources" we'll end up with some code like this:
`framePresentation.isCallFrame() ? framePresentation.callFrame.codeFlow.location.position.lineNumber :  framePresentation.codeFlow.location.position.lineNumber
`
instead of just writing: 
`framePresentation.lineNumber
`

So I'd definitely love to keep the duplicated info...

Reply:
Ehhhhh ok